### DISCUSSION TITLE
Linting fails when clippy finds an error in a dependency

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [mrcjkb/rustaceanvim#626] [Q&A] [answered]
### Have you read the docs and searched existing issues?

- [X] I have read [the README's troubleshooting guide](https://github.com/mrcjkb/rustaceanvim/blob/master/README.md#stethoscope-troubleshooting) and FAQ(https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#left_speech_bubble-faq).
- [X] I am using the latest rustaceanvim release.
- [X] I have searched [existing issues](https://github.com/mrcjkb/rustaceanvim/issues?q=is%3Aopen+is%3Aissue+label%3Abug).
- [X] I have searched [existing discussions](https://github.com/mrcjkb/rustaceanvim/discussions).
- [X] I have searched the existing issues of other plugins related to this issue (if applicable).
- [X] I have searched the [existing rust-analyzer issues](https://github.com/rust-lang/rust-analyzer/issues?q=is%3Aissue+) (if applicable).
- [X] I can reproduce the issue with stable Neovim (not a distribution or nightly) and the [minimal config](https://github.com/mrcjkb/rustaceanvim/blob/master/troubleshooting/minimal.lua).
- [X] This is an issue about rustaceanvim, not rust-analyzer or Neovim.
- [X] This is [not an issue with autocompletion](https://github.com/mrcjkb/rustaceanvim?tab=readme-ov-file#im-having-issues-with-autocompletion).

### Neovim version (nvim -v)

v0.10.2

### Operating system/version

NixOS unstable

### Output of :checkhealth rustaceanvim

```console
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- WARNING dap not installed. Needed for debugging features [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap)

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 1.81.0-nightly (6292b2a 2024-07-02)
- OK Cargo: found cargo 1.81.0-nightly (4ed7bee47 2024-06-25)
- OK rustc: found rustc 1.81.0-nightly (6292b2af6 2024-07-02)
- ERROR Failed to run healthcheck for "rustaceanvim" plugin. Exception:
  .../data/nvim/lazy/rustaceanvim/lua/rustaceanvim/health.lua:104:       cargo-nextest: not found: Could not find an executable binary.
        rustaceanvim requires [cargo-nextest](https://nexte.st).
              Next generation test runner for Rust projects.
        Optional dependency, required if the 'tools.enable_nextest' option is set.
```

My actual config:
```
rustaceanvim: require("rustaceanvim.health").check()

Checking for Lua dependencies ~
- OK [mfussenegger/nvim-dap](https://github.com/mfussenegger/nvim-dap) installed.

Checking external dependencies ~
- OK rust-analyzer: found rust-analyzer 1.81.0-nightly (6292b2a 2024-07-02)
- OK Cargo: found cargo 1.81.0-nightly (4ed7bee47 2024-06-25)
- OK rustc: found rustc 1.81.0-nightly (6292b2af6 2024-07-02)
- OK debug adapter: found codelldb 

Checking config ~
- OK No errors found in config.

Checking for conflicting plugins ~
- OK No conflicting plugins detected.

Checking for tree-sitter parser ~
- OK tree-sitter parser for Rust detected.
```

### How to reproduce the issue

`Cargo.toml`:
```toml
[workspace]
members = [ "mylib","t1"]
```

`./mylib/Cargo.toml`:
```toml
[package]
name = "mylib"
version = "0.1.0"
edition = "2021"

[dependencies]
```

`./mylib/src/lib.rs`:
```rs
#![deny(unused)]
pub fn add(left: u64, right: u64) -> u64 {
    let a = 3;
    left + right
}
```

`./t1/Cargo.toml`:
```toml
[package]
name = "t1"
version = "0.1.0"
edition = "2021"

[dependencies]
mylib = { path = "../mylib" }
```

`./t1/src/main.rs`:
```rs
fn main() {
    let a: i32 = "";
    println!("Hello, world!");
}
```

Now open `./t1/src/main.rs` with `nvim` (or `nvim -u minimal.lua`). The erroneous line is not linted. Sanity check: open `./mylib/src/lib.rs`, it lints the unused variable.

I believe it has to do with `rustaceanvim` because it works both with lspconfig's and rust-tools configurations of RA, but not with `rustaceanvim`.

### Expected behaviour

It lints both the unused variable in the dependency AND a type mismatch in the bin project.
![image](https://github.com/user-attachments/assets/2a88bd6b-0b61-46ce-8f57-051f18b97e42)

### Actual behaviour

If there's a clippy error in a dependency, the linting stops there.

![image](https://github.com/user-attachments/assets/495505f6-f20d-4f22-b8c5-34002a2ff731)

It's worth noting that it works as usual if it is not an error, but, for example, a warning:
![image](https://github.com/user-attachments/assets/d72cf5aa-07be-4223-91b7-953aef0a4c57)

### The minimal config used to reproduce this issue.

```Lua
The one suggested by this form reproduces the issue:
https://github.com/mrcjkb/rustaceanvim/blob/master/troubleshooting/minimal.lua
```

**Comment by mrcjkb** [comment]
Hey 👋 

> I believe it has to do with rustaceanvim because it works both with lspconfig's and rust-tools configurations of RA, but not with rustaceanvim.

This is because rustaceanvim enables clippy by default, if it detects that clippy is installed. 
If you don't want that behaviour, you can disable clippy, see `:h rustaceanvim.config`.

If you want to use clippy with rust-analyzer, then I suggest opening an issue there.

**Reply by WhiteBlackGoose** [reply]
Hi, thanks for the quick response!

So I tried disabling it:
```lua
vim.g.rustaceanvim = {
  tools = {
    ...
    enable_clippy = false,
  },
  ...
```

now if I open the file directly, e. g. `nvim t1/src/main.rs`, linting works.

But if I go to `mylib/src/main.rs`, I will see no linting. However linting appears if I save the buffer. But in that case, it disappears from `t1/src/main.rs` (and even saving doesn't restore it)

P. S. ok I see that the default RA configuration works the same way. Hmmm... how did rust-tools solve it then. Rust-tools worked really well in that regard :thinking: 

P. P. S. Ahhh I had RA with rust-tools, but now I disabled it. 
> If you want to use clippy with rust-analyzer, then I suggest opening an issue there

Where? I'd like it to work the same way rust-tools + RA did. I need both to provide linting, I guess

**Reply by WhiteBlackGoose** [reply]
I can't find a way around it for now. Really sad, maybe I'll have to migrate back to rust-tools. Thanks for your help

**Reply by mrcjkb** [reply]
I just tried to reproduce using your setup, and made the following observation:

- If I comment out the `#![deny(unused)]` from `mylib/src/lib.rs`, linting works fine in both packages.
- If I introduce an error (it doesn't matter if it's a clippy error, e.g. change `left: u64` to `left: &str`), it stops linting in `t1/src/main.rs`.

This appears to be because `t1` depends on `mylib` and linting will fail if `mylib` has a compile error.

I don't know why it worked for you with rust-tools, but this is likely caused by the toolchain (you're either using a different version or have it configured differently with rustaceanvim). Rustaceanvim doesn't explicitly set any configs related to this, so it's more likely the toolchain.

**Reply by WhiteBlackGoose** [reply]
> made the following observation

Yes, I reflected exactly in my screenshots. 

>  don't know why it worked for you with rust-tool

I guess because rust-tools used clippy and I enabled RA on top via lspconfig, so they both worked and yielded different linting, which together gave me what I needed. Now rustaceanvim discourages doing that, and it does work weird so I'm not going to hack it, but maybe there's another way..

**Reply by mrcjkb** [reply, marked as answer]
What do you mean by " enabled RA on top via lspconfig"? RA = rust-analyzer?
rustaceanvim uses rust-analyzer, too. It just doesn't go through nvim-lspconfig. Both rustaceanvim and nvim-lspconfig use `vim.lsp.start` to start the client. And as I said, rustaceanvim doesn't do anything to configure diagnostic settings.

If you're absolutely sure you're using the same toolchain with both plugins, I suggest you run `:lua =vim.lsp.get_clients()[1].settings` and `:lua =vim.lsp.get_clients()[1].capabilities` and compare between the two plugins.
